documentation.md

```diff
@@ -0,0 +1,97 @@
+
+### Model Layer
+
+Throughout these docs we'll be using these two basic models for a chat app, similar to the
+[Hotwire demo video](https://www.youtube.com/watch?v=eKY-QES1XQQ):
+
+```python
+# chat/models.py
+
+from django.db import models
+class Room(models.Model):
+    name = models.CharField(max_length=255)
+
+
+class Message(models.Model):
+    room = models.ForeignKey(Room, related_name="messages", on_delete=models.CASCADE)
+    text = models.TextField()
+```
+
+`turbo-django` provides a `BroadcastableMixin` that enables CRUD operations from a given model
+to be broadcast over websockets. To enable messages to be broadcast over websocket to users
+listening in on a particular room, we add the `BroadcastableMixin` and tell `turbo-django`
+that actions on messages should be broadcast to the related `Room` as well:
+
+```python
+class Message(BroadcastableMixin, models.Model):
+    broadcast_to = ["room"]
+    broadcast_self = False
+
+    room = models.ForeignKey(Room, related_name="messages", on_delete=models.CASCADE)
+    text = models.TextField()
+    created_at = models.DateTimeField(auto_now_add=True)
+```
+
+`broadcast_to` can contain any foreign key field on the model. We set `broadcast_self` to `False` since we don't
+anticipate listening in to updates on a single message, only on the associated `Room`. This is all the set-up on the
+model side of things that we'll need!
+
+### Template Layer
+
+#### Subscribing to a stream
+Next, we can declare that a given page should subscribe to broadcasts for a given model instance. Make sure to
+`{% load turbo_streams_helpers %}` to get access to the template tag in your template.
+
+Make sure to pass the model instance (here, it's passed as `room`) into the template context. Somewhere in the `<body>`,
+we need to add the template tag `{% turbo_stream_from room %}`. That's all we need to do to connect this template, when
+rendered on the front-end, to the stream of updates for a given model instance!
+
+#### Rendering Streams
+Streams need to render HTML of updates on the backend using template partials to send over to the frontend over the wire.
+`turbo-django` at the moment relies on some conventional names for templates and IDs:
+
+- `app_name/model_name.html` should contain the template partial for rendering the model. The context passed in will
+  have the labels `object` as well as `model_name` refer to the instance being rendered. Streams mandates that there's
+  one, top-level element wrapping the HTML. This element should have an ID of the form `<model_name>_<pk>`, where `pk`
+  is the primary key of the instance. The element's ID is necessary to ensure that the Streams update operation, titled
+  `replace`, can predictably find elements on the page.
+  * For our example with the `Message` model, we should create a template called `chat/message.html`, where the top-level
+    HTML element should have the ID `id={{ message.pk }}`. `object.pk` would work equally well.
```

Comment:
I wonder if the lib should take care of this? Two thoughts:

1. The lib could wrap all templates in a `div` and take care of putting the ID there or
2. The lib could generate the ID for the user and pass `turbo_id` into the context of the template for the user to use.

Either option would put less work on the user (and thus less user error) to make the ID themselves. We could add another, optional, class var on the Broadcast mixin that tells us what field to use for the model in case the default `pk` isn't what they want.

Reply:
I agree it would definitely be an improvement for turbo to take care of this, but the one issue I see is re-usability of the partial. It seems like a big benefit of how streams is designed is that you can re-use your template partials that you use when rendering initial pages and frames. Relying on `turbo_id` being in the context then doesn't seem like the right move to me... maybe a possibility would be to make a `turbo_id` template tag where you pass in the model that works in any context... so you would write `id="{% turbo_id message %}" and the template tag will determine the proper ID based on settings from the broadcast mixin? Kind of like rails's `dom_id`, but that isn't built into django.

Reply:
That's a really good point. Reusing templates is definitely ideal. In fact, the `turbo_id` template tag could just not take any arguments. All it would need to do is optionally pull the ID variable from the context if it exists, otherwise do nothing.